rfcs/20180726-tf-data-windowing-reducers.md

```diff
@@ -0,0 +1,244 @@
+
+  def max_reduce_fn(state, value):
+    return tf.maximum(state, tf.shape(value))
+
+  def max_finalize_fn(state):
+    return state
+
+  # Compute the padded shape.
+  max_reducer = tf.contrib.Reducer(max_init_fn, max_reduce_fn, 
+                                   max_finalize_fn)
+  padded_shape = dataset.reduce(max_reducer)
+
+  def batch_init_fn(_):
+    return tf.fill(
+        tf.concat([np.array([0], dtype=np.int32), padded_shape], 0),
+        tf.constant(padding_value, dtype=dataset.output_types))
+
+  def batch_reduce_fn(state, value):
+    return tf.concat([state, [value]], 0)
+
+  def batch_finalize_fn(state):
+    return state
+
+  def pad_fn(value):
+    shape = tf.shape(value)
+    left = tf.zeros_like(shape)
+    right = padded_shape - shape
+    return tf.pad(value, tf.stack([left, right], 1), 
+                  constant_values=padding_value)
+
+  batch_reducer = tf.data.Reducer(batch_init_fn, batch_reduce_fn, 
+                                  batch_finalize_fn)
+  return dataset.map(pad_fn).reduce(batch_reducer)
+
+padded_batch = padded_batch_dense(
+    tf.data.Dataset.from_tensor_slices([[1], [2]]), [2], 0)
+with tf.Session() as sess:
+  print(sess.run(padded_batch)) # produces [[1 0] [2 0]]
+```
+
+
+Note that the method uses two reducers. The first reducer is used to compute the shape to pad to (as the maximum shape of the input elements) and the second reducer is used to do the actual batching.
+
+
+### End-to-end Example
+
+Bringing it all together, we now illustrate how to combine the window transformation and reducers to perform generalized tf.data batching:
+
+
+```python
+import tensorflow as tf
+
+def gen():
+  yield ('a', [1])
+  yield ('b', [2])
+  yield ('c', [3])
+  yield ('d', [4, 4])
+
+def map_fn(a, b):
+  return batch_dense(a), padded_batch_dense(b, [2], 0)
+  
+dataset = tf.data.Dataset.from_generator(gen, (tf.string, tf.int32))
+dataset = dataset.window(2).map(map_fn)
+get_next = dataset.make_one_shot_iterator().get_next()
+
+with tf.Session() as sess:
+  print(sess.run(get_next)) # produces (['a', 'b'], [[1, 0], [2, 0]])
+  print(sess.run(get_next)) # produces (['c', 'd'], [[3, 0], [4, 4]])
+```
+
+
+
+## API Changes
+
+This design document proposes the following changes to the tf.data API:
+
+*   Adding a `tf.data.Dataset.window` method, which provides the windowing functionality described in this proposal.
+*   Promoting the `tf.contrib.data.reduce_dataset()` method to `tf.data.Dataset.reduce()` and the `tf.contrib.data.Reducer` class to `tf.data.Reducer`.
+*   Allowing nested datasets as inputs of `map` and `filter`.
+*   Adding canned reducers for batching and padded batching of dense and sparse tensors to `tf.contrib.data`.
+
+
+## Open Questions
+
+*   Any interest in the window transformation supporting parameters for specifying the window shift and stride (similar to tf.contrib.data.sliding_window_batch)? Is there any other type of windowing that people are interested in?
```

Comment:
What about the field of video analysis video inpainting? Have we some common shift a stride scenarios?

Reply:
The stride/shift uses cases for video analysis I am aware of are covered by `tf.contrib.data.sliding_window_batch`.